components/movementsensor/dualgps/dualgps.go

```diff
@@ -0,0 +1,236 @@
+	}
+
+	return nil
+}
+
+type gpsWithLock struct {
+	gps     movementsensor.MovementSensor
+	gpsLock sync.Mutex
+}
+
+func (lgps *gpsWithLock) getPosition(ctx context.Context, extra map[string]interface{}) (*geo.Point, error) {
+	lgps.gpsLock.Lock()
+	defer lgps.gpsLock.Unlock()
+	pos, _, err := lgps.gps.Position(ctx, extra)
+	return pos, err
+}
+
+// Position gets the position of a fake movementsensor.
+func (dg *dualGPS) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	return geo.NewPoint(math.NaN(), math.NaN()), math.NaN(), movementsensor.ErrMethodUnimplementedPosition
+}
+
+// getHeading calculates bearing and absolute heading angles given 2 geoPoint coordinates
+// 0 degrees is North, 90 degrees is East, 180 degrees is South, 270 is West.
+func getHeading(first, second *geo.Point, yawOffset float64) (float64, float64, float64) {
+	// convert latitude and longitude readings from degrees to radians
+	// so we can use go's periodic math functions
+	firstLat := utils.DegToRad(first.Lat())
+	firstLong := utils.DegToRad(first.Lng())
+	secondLat := utils.DegToRad(second.Lat())
+	secondLong := utils.DegToRad(second.Lng())
+
+	// calculate the bearing between gps1 and gps2
+	deltaLong := secondLong - firstLong
+	y := math.Sin(deltaLong) * math.Cos(secondLat)
+	x := math.Cos(firstLat)*math.Sin(secondLat) - math.Sin(firstLat)*math.Cos(secondLat)*math.Cos(deltaLong)
+	bearing := utils.RadToDeg(math.Atan2(y, x))
+
+	// maps the bearing to the range of 0-360 degrees
+	if bearing < 0 {
+		bearing += 360
+	}
+
+	// calculate heading from bearing, accounting for yaw offset between the two gps devices
+	// e.g if the MovementSensor antennas are mounted on the left and right sides of the robot,
+	// the yaw offset would be roughly 90 degrees
+	heading := bearing - yawOffset
+	// make heading positive again
+	if heading < 0 {
+		diff := math.Abs(heading)
+		heading = 360 - diff
+	}
+
+	// return the standard bearing in the -180 to 180 range, with North being 0, East being 90
+	// South being 180/-180 and West being -90
+	var standardBearing float64
+	if bearing > 180 {
+		standardBearing = -(360 - bearing)
+	} else {
+		standardBearing = bearing
+	}
+
+	return bearing, heading, standardBearing
+}
+
+// CompassHeading gets the compass headings of a fake movementsensor.
+func (dg *dualGPS) CompassHeading(ctx context.Context, extra map[string]interface{}) (float64, error) {
+	geoPoint1, err := dg.gps1.getPosition(context.Background(), extra)
+	if err != nil {
+		return math.NaN(), err
+	}
+
+	geoPoint2, err := dg.gps2.getPosition(context.Background(), extra)
+	if err != nil {
+		return math.NaN(), err
+	}
+
+	_, heading, _ := getHeading(geoPoint1, geoPoint2, 0)
+	return heading, nil
+}
+
+func (dg *dualGPS) Properties(ctx context.Context, extra map[string]interface{}) (*movementsensor.Properties, error) {
+	return &movementsensor.Properties{
+		CompassHeadingSupported:     true,
+		PositionSupported:           false,
```

Comment:
I'm open to disagreement, but I think position should be trivially supported by just getting the position of the first GPS.

Reply:
rejected in scope.